File: docker_auth.py
"""Docker authentication and pull secret management for the roxie deployment tool."""

import base64
import json
import os
import subprocess
from typing import Any

import yaml
from rich.console import Console

from errors import RoxieError

registry = "quay.io"

class DockerAuth:
    """Handles Docker authentication and pull secret management"""

    def __init__(self, console: Console | None = None, cache_enabled: bool = True):
        self.console = console or Console()
        self.cache_enabled = cache_enabled
        self._auth_cache: dict[str, str] = {}

    def get_encoded_docker_auth(self, username: str | None = None, password: str | None = None) -> str:
        """Generate Docker authentication string for image pull secrets"""
        # Try environment variables first
        env_username = os.environ.get("REGISTRY_USERNAME")
        env_password = os.environ.get("REGISTRY_PASSWORD")

        if env_username and env_password:
            username = env_username
            password = env_password
            auth_string = f"{username}:{password}"
            return base64.b64encode(auth_string.encode()).decode()

        docker_config_path = os.path.expanduser("~/.docker/config.json")
        if os.path.exists(docker_config_path):
            encoded_auth = self.get_registry_auth_from_config(docker_config_path, registry)
            if encoded_auth:
                return encoded_auth

        raise RoxieError("No Docker credentials found")

    def get_cred_helper_from_config(self, config_path: str, registry: str) -> str:
        """Get cred helper from existing Docker config"""
        with open(config_path) as f:
            docker_config = json.load(f)

        cred_helper = ""
        cred_store = docker_config.get("credStore", "")
        if cred_store and isinstance(cred_store, str):
            cred_helper = cred_store
        cred_helpers = docker_config.get("credHelpers", {})
        if registry in cred_helpers and isinstance(cred_helpers[registry], str):
            cred_helper = cred_helpers[registry]

        return cred_helper

    def format_cred_helper_result(self, result: dict[str,str]) -> str:
        """Format Docker cred helper result"""
        return f"{result['Username']}:{result['Secret']}"

    def invoke_cred_helper(self, cred_helper: str, registry: str) -> dict[str,Any]:
        """Invoke Docker cred helper"""
        result = subprocess.run(
            [f"docker-credential-{cred_helper}", "get"], input=registry, text=True, capture_output=True)
        stdout = result.stdout.strip()
        json_result = json.loads(stdout)
        if not isinstance(json_result, dict):
            raise RoxieError(f"Invalid cred helper result: {stdout}")
        return json_result

    def get_registry_auth_from_config(self, config_path: str, registry: str) -> str:
        """Extract encoded auth from existing Docker config"""
        with open(config_path) as f:
            docker_config = json.load(f)

        auths = docker_config.get("auths", {})

        if registry not in auths:
            return ""

        if auths[registry] == {}:
            cred_helper = self.get_cred_helper_from_config(docker_config, registry)
            if cred_helper:
                helper_result = self.invoke_cred_helper(cred_helper, registry)
                return self.format_cred_helper_result(helper_result)

        if isinstance(auths[registry], str):
            return auths[registry]

        return ""

    def create_pull_secret_yaml(self, namespace: str) -> str:
        """Create Kubernetes pull secret YAML"""
        encoded_auth = self.get_encoded_docker_auth()
        docker_config = {"auths": {registry: {"auth": encoded_auth}}}
        encoded_config = base64.b64encode(json.dumps(docker_config)).encode()
        secret = {
            "apiVersion": "v1",
            "kind": "Secret",
            "metadata": {
                "name": "stackrox",
                "namespace": namespace,
            },
            "type": "kubernetes.io/dockerconfigjson",
            "data": {
                ".dockerconfigjson": encoded_config,
            },
        }
        return yaml.dump(secret)

File: pkg/dockerauth/dockerauth.go
package dockerauth

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"

	"github.com/stackrox/roxie/pkg/logger"
)

const (
	acsImageRegistry = "quay.io"
)

// DockerAuth handles Docker authentication and pull secret management.
type DockerAuth struct {
	logger *logger.Logger
}

// DockerConfig represents Docker configuration structure.
type DockerConfig struct {
	Auths       map[string]AuthEntry `json:"auths,omitempty"`
	CredHelpers map[string]string    `json:"credHelpers,omitempty"`
	CredsStore  string               `json:"credsStore,omitempty"`
}

// AuthEntry represents a single auth entry in Docker config.
type AuthEntry struct {
	Auth string `json:"auth,omitempty"`
}

// CredentialData represents credential data from credential helper.
type CredentialData struct {
	Username string `json:"Username"`
	Secret   string `json:"Secret"`
}

// New creates a new DockerAuth instance.
func New(log *logger.Logger) *DockerAuth {
	return &DockerAuth{
		logger: log,
	}
}

// GetDockerAuthString generates Docker authentication string for image pull secrets
func (d *DockerAuth) GetDockerAuthString(_, _ string) (string, error) {
	var username, password string

	// Try environment variables first.
	username = os.Getenv("REGISTRY_USERNAME")
	password = os.Getenv("REGISTRY_PASSWORD")

	if username != "" && password == "" {
		return "", errors.New("REGISTRY_USERNAME set but REGISTRY_PASSWORD is empty")
	}
	if username == "" && password != "" {
		return "", errors.New("REGISTRY_PASSWORD set but REGISTRY_USERNAME is empty")
	}

	if username == "" {
		// Try to get from Docker config file.
		dockerConfigPath := filepath.Join(os.Getenv("HOME"), ".docker", "config.json")
		d.logger.Dimf("REGISTRY_USERNAME/REGISTRY_PASSWORD unset. Trying to obtain Docker credentials from config file: %s", dockerConfigPath)
		if _, err := os.Stat(dockerConfigPath); err == nil {
			var err error
			username, password, err = d.getCredentialsFromDockerConfig(dockerConfigPath)
			if err != nil {
				return "", err
			}
		}
	}

	if username == "" || password == "" {
		return "", errors.New("no Docker credentials found")
	}

	// Create auth string.
	authString := fmt.Sprintf("%s:%s", username, password)
	encodedAuth := base64.StdEncoding.EncodeToString([]byte(authString))

	dockerConfig := DockerConfig{
		Auths: map[string]AuthEntry{
			acsImageRegistry: {Auth: encodedAuth},
		},
	}

	jsonData, err := json.Marshal(dockerConfig)
	if err != nil {
		return "", fmt.Errorf("failed to marshal Docker config: %w", err)
	}

	return string(jsonData), nil
}

// getCredentialsFromDockerConfig extracts credentials from existing Docker config.
func (d *DockerAuth) getCredentialsFromDockerConfig(configPath string) (string, string, error) {
	data, err := os.ReadFile(configPath)
	if err != nil {
		return "", "", fmt.Errorf("failed to read Docker config: %w", err)
	}

	var config DockerConfig
	if err := json.Unmarshal(data, &config); err != nil {
		return "", "", fmt.Errorf("failed to parse Docker config: %w", err)
	}

	// Check for existing auths for the ACS image registry.
	if authEntry, ok := config.Auths[acsImageRegistry]; ok && authEntry.Auth != "" {
		// Decode the base64 auth string to get username:password
		decoded, err := base64.StdEncoding.DecodeString(authEntry.Auth)
		if err != nil {
			return "", "", fmt.Errorf("failed to decode auth string: %w", err)
		}
		parts := bytes.SplitN(decoded, []byte(":"), 2)
		if len(parts) != 2 {
			return "", "", errors.New("invalid auth format")
		}
		return string(parts[0]), string(parts[1]), nil
	}

	// Try credential helper specifically configured for the ACS image registry
	helper := d.lookupCredentialHelperForRegistry(&config, acsImageRegistry)
	if helper == "" {
		return "", "", fmt.Errorf("no Docker credentials found in config for ACS image registry (%s)", acsImageRegistry)
	}

	credData, err := d.getCredentialFromHelper(helper, acsImageRegistry)
	if err != nil {
		return "", "", fmt.Errorf("failed to get credentials from helper '%s' for '%s': %w", helper, acsImageRegistry, err)
	}

	return credData.Username, credData.Secret, nil
}

// lookupCredentialHelperForRegistry returns the credential helper name for a given registry
// by checking registry-specific credHelpers first, then falling back to the global credsStore.
// Returns empty string if no helper is configured.
func (d *DockerAuth) lookupCredentialHelperForRegistry(config *DockerConfig, registry string) string {
	// First check for registry-specific credential helper
	if helper, ok := config.CredHelpers[registry]; ok {
		return helper
	}

	// Fall back to global credential store
	return config.CredsStore
}

// getCredentialFromHelper retrieves credentials from a credential helper.
func (d *DockerAuth) getCredentialFromHelper(helperName, registry string) (*CredentialData, error) {
	cmd := exec.Command(fmt.Sprintf("docker-credential-%s", helperName), "get")
	cmd.Stdin = bytes.NewBufferString(registry)

	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("credential helper '%s' for '%s' failed: %w", helperName, registry, err)
	}

	var credData CredentialData
	if err := json.Unmarshal(output, &credData); err != nil {
		return nil, fmt.Errorf("failed to parse credential helper output: %w", err)
	}

	if credData.Username == "" || credData.Secret == "" {
		return nil, errors.New("credential helper returned empty credentials")
	}

	return &credData, nil
}

// CreatePullSecretYAML creates Kubernetes pull secret YAML.
func (d *DockerAuth) CreatePullSecretYAML(namespace string) (string, error) {
	dockerConfigJSON, err := d.GetDockerAuthString("", "")
	if err != nil {
		return "", err
	}

	encodedConfig := base64.StdEncoding.EncodeToString([]byte(dockerConfigJSON))

	secretYAML := fmt.Sprintf(`apiVersion: v1
kind: Secret
metadata:
  name: stackrox
  namespace: %s
type: kubernetes.io/dockerconfigjson
data:
  .dockerconfigjson: %s
`, namespace, encodedConfig)

	return secretYAML, nil
}
